draft: Add proposal for grafana dashboard plugin (#2708)
kubernetes-sigs/kubebuilder: designs/operator-monitoring-plugin.md

---
title: controller-observation-plugin
authors:
  - "@Kavinjsir"
reviewers:
  - "@varsha"
  - "@camila"
approvers:
creation-date: 2022-06-06
last-updated: 2022-06-06
status: provisional
see-also:
  - ""
replaces:
  - ""
superseded-by:
  - ""
---

# a-grafana-dashboard-plugin-to-observe-controller-metrics

## Release Signoff Checklist

- [ ] Enhancement is `implementable`
- [ ] Design details are appropriately documented from clear requirements
- [ ] Test plan is defined
- [ ] Graduation criteria for dev preview, tech preview, GA

## Open Questions [optional]

> - TBD

Reviewer: WDYT about : 
a) Should we scaffold a dir with the config to install the grafana? See: https://grafana.com/docs/grafana/latest/installation/kubernetes/

b) What should be the name of the default directory and where it should be scaffold?
It seems that we could scaffold: (root dir) something like:

grafana/
grafana/dashboards
grafana/config

OR 

inside of the config dir another grafana dir with the k8s manifests to be applied on the cluster when the plugin is used

Reviewer 2: We can also add a few other pointers like:
1. How is the plugin backwards compatible for users who have prometheus scaffolded?
- Is it compatible only with specific versions of prometheus? Or do can we have an existing kb project and add this plugin to scaffold grafana manifests smoothly. Do we need any other changes to an old project structure.

Author: That's great! I like all these questions, they helps exploring the project.  Thx @camilamacedo86 @varshaprasad96

## Summary

> This EP will provide a way for the operator author to visualize the runtime metrics in order to observe the operator status.

Reviewer 2: ```suggestion
This EP will provide a way for the operator author to visualize the runtime metrics in order to observe the operator status using Grafana.
```

The functionality will be implemented by a new plugin that can generate the 'manifests' to be loaded on Grafana.

> In the context of this EP, the Grafana 'manifests' are mainly:

Reviewer 2: How about elaborating this a bit more on why plugin will add value. Something on the lines of - we currently have controller-runtime metrics exposed by default for all KB operators at an endpoint through Prometheus. In order to visualize those same metrics, every operator author has to individually put in the effort to create a dashboard. This plugin intends to reduce the effort of operator authors to create dashboards and instead provides with the manifests that will allow them to integrate their controller metrics with grafana easily. On top of the basic dashboard provided by us having default metrics, operator authors can also add custom metrics and visualize them  in the format they are comfortable in.

- JSON files to be directly loaded on Grafana dashboards

> - TBD

Reviewer 2: Will the grafana dashboards utilize any of the existing promethtues manifests?

## Motivation

The [controller-runtime](https://github.com/kubernetes-sigs/controller-runtime) provides rich metrics to measure controllers' status from various perspectives.

Currently, users of Kubebuilder need to design and create their own Grafana dashboards to visualize the metrics.
It would be great if we can provide a unified and comprehensive approach so that users may:

- easily visualize controller/operator metrics
- avoid duplicate creation of Grafana dashboards file
- improve the observability towards the opeartor, keep it safe and robust

### Goals

- Provide manifests that can display operator status for common needs.
- Add a plugin or enhance the current kustomize plugin that scaffolds Grafana manifests.

### Non-Goals

> - TBD

Reviewer 2: Some non-goals could be:
1. We do not deal with the kind of metrics which will be exported.
2. The proposal will talk not only about scaffolding basic manifests (or templates) for integrating grafana dashboards. It does not deal with dashboard designs.

## Proposal

This proposed implementation is to create a new plugin, `grafana`, that scaffolds the manifests which are loaded by Grafana to display dashboards for controllers' status.

> Basically, the manifest can be raw json file such as `controller-dashboard.json` that can be directly applied to Grafana Web UI.

Reviewer: Is required to be a JSON file OR could we have the manifest with the dashboard?
So that, users would only apply the manifests by command line and check the dashboards with graphs and end etc on the UI directly? 

c/c @Kavinjsir @varshaprasad96

Reviewer 2: The JSON file provides the [dashboard](https://grafana.com/docs/grafana/latest/dashboards/json-model/) template, (ie) UI related. How do we apply it on cluster? Could we elaborate on that? In the sense, do the users have to just import this JSON specified dashboard in their grafana instance or is there any other way to do so?

Author: For this json file, the basic approach is to let the user **copy** the content and **paste** it in their Grafana Web UI.

So this would assume:
1. The users have the write access to their grafana service
2. The users' Grafana service allows dashboards creation and storage from Web UI.

Author: The above idea may not be ideal. It assumes users are commonly doing dashboards creation work on Web UI directly.

I know for `prometheus-operator` (or `kube-prometheus`), the json content can be written in a `K8s ConfigMap`. That means:
- The dashboard is installed when the configmap is applied to the cluster.
- The installation process can be handled in Makefile with a simple command.

I personally like this approach, but I'm not sure how common is `prometheus-operator` or `kube-prometheus`.

Or can we assume people would use these stacks?

Author: In other case, like some people who are familiar with the Grafana community, they may setup their monitoring stacks through jsonnet. The grafana community provides [tanka](https://tanka.dev/) to provision k8s resources that can operate prometheus and grafana.

In that case, it maybe hard to install the dashboard at our side. So this maybe another story.

Author: So it seems challenging to provide a general approach to fit all cases. The dashboard installation process seems depends on how users deploy their Grafana.

This is also the reason why I come back to the very basic way as just scaffolding out the raw json file.

Author: Would you have some ideas or suggestions on this?

As there are many useful metrics, it might be better to observe in multiple dashboards rather than one.

Hence, we may also provide optional dashboards that focus on different aspects.

For instance: work queue, reconciliation, performance, webhook.

> As a result, the layout given by the plugin can be a folder of `config/grafana/`:

Reviewer: In the config/ dir we have only manifests to be used with kustomize. See that SDK add new features by adding manifests there, i.e for scorecard (https://github.com/operator-framework/operator-sdk/tree/master/testdata/go/v3/memcached-operator/config/scorecard) however, all is using kustomize, so I am afraid that we should not add JSON files on the config/ dir. However, we can create a new directory in the root like grafana/ where we can add all there.

Author: Yeah, this sounds better!

```shell
config/grafana
├── controller-dashboard.json # The main dashboard raw json file for the simplest usage(copy/paste)
├── options
│  ├── performance.json       # CPU/Memory usaage, restful api performance
│  ├── reconcilier.json       # Reconciliation instruments
│  ├── webhook.jon            # Webhook Ops
│  └── workqueue.json         # Work queue status

> └── README.md                 # Docs for the usage

Reviewer 2: The specific `.json` file to refer to each metric, right? What happens when that metric is unavailable for a controller. Do we just get a blank dashbaord with no metrics? Also would we be hardcoding the all the metrics that controller-runtime exports? If so, we should probably start by scaffolding only a subset which are really important/widely used.

> └── README.md                 # Docs for the usage

Author: Yeah, @varshaprasad96 's idea is mainly what I thought. Since the controller-runtime provides plenty metrics, it seems not necessary/good to display all of them in one dashboard.

In brief when the queries failed (controller down/authentication failure/syntax error), the panels will be blank. By that time, the dashboard itself is added into the Grafana service, the user may get some error notification when open the page.

I am planning to hardcoding. Since we hope to have the dashboard available on the Grafana market, that means a stable json file, which may be likely as a statice file maintained in the plugin implementation folder?

How would you think of this? Would you have some suggestions or ideas to deal with this more elegantly?

```

The plugin is triggered to do the scaffolding work for the layout above when the following commands are executed:

- `kubebuilder init -–plugins=grafana/alpha-v1`: when initializing a new project
- `kubebuilder edit -–plugins=grafana/alpha-v1`: when adding features to an existing project

#### Dashboard Panels

The creation of the panels inside the dashboard can be referred to default [controller-runtime metrics](https://github.com/kubernetes-sigs/kubebuilder/blob/master/docs/book/src/reference/metrics-reference.md).

The overall dashboard should reflect:

1. Latency: The duration of a request for certain controllers/operators, which includes both successful ops and errors ops.
2. Traffic: The measure of how busy the controllers/operators can be.
3. Errors: Error rates, or the operations/requests that takes unexpectedly long time.
4. Saturation: The measure of the possibility of overflows such as deep work queue depth, concurrencies of reconciliations, large cpu/memory usage.

> Also, the dashboard should provide filters and selectors to focus on certain objects. This can be implemented by utilizing [Grafana Query Variables](https://grafana.com/docs/grafana/latest/datasources/prometheus/#query-variable).

Reviewer 2: These queries will be specified in JSON files, or the user needs to filter them manually in a dashboard?

An example can be:

> ![sample-dash](https://doc-0c-bg-docs.googleusercontent.com/docs/securesc/op3r6mi7tplp3i74fcm5ur03mtb42v46/84bkk8a1kontsl4d7ehvrkgs1920s6j5/1654557000000/13713818693191747359/13713818693191747359/1F0eM6qYGCiZZ3dqYOKUG11qxqFSDg9kN?e=view&ax=ACxEAsbPfXom32Dkn-LdACjf86ktIPBQRx3OYuF5l9aSUjEYKcraRYHab1nPuy1VXgDu070T-Qe1VcWWa8ItghTzaP8m39BQJ1ATqVgbdPRlLGcjA8vnjTt_OiRekntAPtedbfLEmCNTSCEDuWK0UBGnln4WbKxiAf7WyOUCRqiy47IWRpUwYXgFIfz4kH6vDoVKaWj8CkFFAKWxnbUz22uQayoK2jwgnFDcB_2EnYiXn8c85tMi9_X9goNyQW64Zi1phrlFvnLlJq6x_bHOdtWZ7mirnoRbu3vmgE0zQezA8rCV1wzTTsfE-161UAdDAawjwaKCFDlyibS-4FTHT0GzZIXu-3ehHx1Txc5YyHpUOnadeSIy6G1qGGY5lzDOk7eGJcInn5vSz0c35Go-PBCI5VoyhljtFh9Irr-VhlIYqwsLcbJWPD9o-Zfmq4-nnPNKb-8_zInFsLjAs4Pj2cLMVDBNeGW7CpOyh4oSrtQSMatyTzdYi4ykAG6aW7KHGoUGpsouhwrQ_WaQbModMmm9742VpyVek-4eB-a4-dSlDXqgUYUHAE4jgvHT5ryeCq7qiLnlNjThafot1o0KaxrBJD4DP1S_BOgN5hXQGKcHiSdi0C_90cIFy9JPVRqlx7zEI4FJDpVBRCs6mECKPt2zR_9veb3wUPTbhkO7g7S81YkcfWkM2gzQDXh6E9ybAqxkFcl_Rk_IvbJ8fvrPUUqUuBuwut7EvQAqlPLqKFoLw39ftcBl2BI&authuser=0&nonce=sk8c49m9ttbto&user=13713818693191747359&hash=6ssipjldoqp2pf2qftgot8fec96m9jbu)

Author: ![image](https://user-images.githubusercontent.com/18136486/172271785-d3fa5a59-83d9-4088-8947-de0274f978a1.png)

Reviewer: I loved the dashboard, my question here is do we need to build them manually by importing the json content OR have we a way to generate the whole content and just apply that to check its result?

cc/ @Kavinjsir @varshaprasad96

Author: Regardless of how it provides the way for user to 'install' dashboards (or auto-installation without manual process), my origin idea is:

Basically, the plugin will scaffold out such json file from static code, which is similar as how `service-monitor` manifest gets generated by [kustomize](https://github.com/kubernetes-sigs/kubebuilder/blob/v3.4.1/pkg/plugins/common/kustomize/v1/scaffolds/internal/templates/config/prometheus/monitor.go#L43).

Which means:
1. We will have a stable content of the json file which may get stored by a constant string.
2. The plugin acts similarly as the `kustomize` plugin, which will generate these json file (considered as grafana manifests).
3. To visualize all default metrics, that may lead to multiple constant strings to store these json content.

Author: I am not sure if there are some approaches to generate the manifests pragmatically. (Compared with the above static way)

And it seems like drawing panels in the Web UI remains the main methods to build dashboards.

(Or if there are better ways to deal with this, I'd be glad to apply.)

Reviewer: HI @Kavinjsir and @varshaprasad96, 

I could check it out. See that with Grafana we can create a config and load the JSON file on it so that we would be able to make a target which when the user executes it the dashboard would be created. 

See the configmap: https://rancher.com/docs/rancher/v2.5/en/monitoring-alerting/guides/persist-grafana/#2-create-a-configmap-using-the-grafana-json-model

That would mean the target would be like: `kubectlc create configmap <dashboard-name> --from-file=/path/to/json/file -n <operator-ns>` OR better yet we scaffold the configMap and we just apply it on the same ns of the Operator. 

So that means:
- The grafana plugin would scaffold the JSON as described here
- Also, it would add a new target that allows the users just run that to create the Dashboard
-The first init PR to add plugin we can only create the JSON file and we can add the target and improve it incrementally in the follow-ups

ASAP I will try to provide further info and suggestions for this EP. 
But you also can see the : https://www.redhat.com/fr/blog/custom-grafana-dashboards-red-hat-openshift-container-platform-4

Reviewer: See here : https://www.lisenet.com/2021/install-and-configure-grafana-on-kubernetes/

@Kavinjsir sharing for you to be able to do manual tests as well and see how we need to config.
I need a little time to be able to finish the suggestions on this doc but you can also apply what was recommended already so that we can close a few comments

├── options
│  ├── performance.json       # CPU/Memory usaage, restful api performance
│  ├── reconcilier.json       # Reconciliation instruments
│  ├── webhook.jon            # Webhook Ops
│  └── workqueue.json         # Work queue status
└── README.md                 # Docs for the usage
```

#### Phase 3 (TBD)

This can be an alternative of **Phase 2** if multiple dashboards are considered not necessary or in a lower priority.

Since metrics exposure and visualization are closely relative, the plugin can be extended to scaffoled the prometheus manifests.
That way, this plugin won't have dependencies on other plugins, which makes it more flexible.

A simple approach maybe:
Provide the `service-monitor` that `kustomize` does.

There are two challenges:

1. How should the plugin behave when it is used with/without `kustomize` being triggered? Would there be any duplications or coverings on the same path of the layout? Or if it maybe some contridictions when using both of them?
2. The current metrics are exposed with restriction so that when the user want to query these metrics on his/her Grafana Dashboard, will it be possible to have authentication/authorization issue?

> 

Author: This might need more discussion to figure out whether necessary and how.

#### Phase 4 (TBD)

This is optional and is necessary to be determined by the community.

Manual installation of Grafana dashboards maybe inconvenient. In some cases, it is possible to make the process automatically when deploying the operator.

However, this may heavily depends on users' stacks of using Grafana:

1. For [prometheus-operator](https://prometheus-operator.dev/docs/developing-prometheus-rules-and-grafana-dashboards/), it supports K8s ConfigMap for dashboards installation.
2. For grafana/jsonnet, users may use [different stack base](https://github.com/grafana/jsonnet-libs/blob/master/grafana/example/tanka/environments/default/dashboards/nyc.json).
3. For Grafana Cloud, users may copy raw json to the ui.

To handle it generally, we may consider [Grafana HTTP API](https://grafana.com/docs/grafana/latest/http_api/dashboard/#create--update-dashboard).

> 

Author: This might need more discussion to figure out whether necessary and how.

status: provisional
see-also:
  - ""
replaces:
  - ""
superseded-by:
  - ""
---

# a-grafana-dashboard-plugin-to-observe-controller-metrics

## Release Signoff Checklist

- [ ] Enhancement is `implementable`
- [ ] Design details are appropriately documented from clear requirements
- [ ] Test plan is defined
- [ ] Graduation criteria for dev preview, tech preview, GA

## Open Questions [optional]

- 1. What does the ideal approach for the plugin to create the grafana manifests?
- 2. Should we scaffold a dir with the config to install the grafana? See: https://grafana.com/docs/grafana/latest/installation/kubernetes/
- 3. Where should the default directory be scaffold?
  - grafana/
  - grafana/dashboards
  - grafana/config
  - OR inside of the config dir another grafana dir with the k8s manifests to be applied on the cluster when the plugin is used
- 4. How is the plugin backwards compatible for users who have prometheus scaffolded?
  - Is it compatible only with specific versions of prometheus? Or do can we have an existing kb project and add this plugin to scaffold grafana manifests smoothly. Do we need any other changes to an old project structure.

## Summary

> This EP will provide a way for the operator author to visualize the runtime metrics in order to observe the operator status.

Reviewer: ```suggestion
This proposal aims to provide an optional helper for the operator author to visualize the runtime metrics which are exported by default in any project that is based on [controller runtime]() and enable the metrics endpoint to observe the operator status using Grafana.
```

Reviewer: Varsha's suggestion + a little nits

Reviewer: Also, see the suggestion in the Motivation section where we try to write down the value that it brings.

We currently have controller-runtime metrics exposed by default for all KB operators at an endpoint through Prometheus.

In order to visualize those same metrics, every operator author has to individually put in the effort to create a dashboard.

This plugin intends to reduce the effort of operator authors to create dashboards and instead provides with the manifests that will allow them to integrate their controller metrics with grafana easily.

On top of the basic dashboard provided by us having default metrics, operator authors can also add custom metrics and visualize them in the format they are comfortable in.

The functionality will be implemented by a new plugin that can generate the 'manifests' to be loaded on Grafana.

> In the context of this EP, the Grafana 'manifests' are mainly:

Reviewer: I would add these parts on the technical details, I think the summary like https://github.com/kubernetes-sigs/kubebuilder/pull/2708/files#r892017769 shows its value as @varshaprasad96 suggested and is a good overview.

> - JSON files to be directly loaded on Grafana dashboards

Reviewer: It is about how it will be done so IHMO also fits in the technical details

- TBD

## Motivation

> The [controller-runtime](https://github.com/kubernetes-sigs/controller-runtime) provides rich metrics to measure controllers' status from various perspectives.
> 
> Currently, users of Kubebuilder need to design and create their own Grafana dashboards to visualize the metrics.
> It would be great if we can provide a unified and comprehensive approach so that users may:
> 
> - easily visualize controller/operator metrics
> - avoid duplicate creation of Grafana dashboards file
> - improve the observability towards the opeartor, keep it safe and robust

Reviewer: WDYT about 
```suggestion

Kubebuilder and SDK users are looking for how to monitor their solutions using Grafana and the Prometheus metrics. Therefore, they are looking for helpers to speed up their solutions and to know how they can do it for their own customized metrics. For further information check some issues raised:

- [Grafana dashboard for visualizing controller-runtime metrics #2183](https://github.com/kubernetes-sigs/kubebuilder/issues/2183)
- [Custom resource metrics #2286](https://github.com/kubernetes-sigs/kubebuilder/issues/2286)

Therefore for SDK users and for those that are looking to distribute their solution via [OLM](https://olm.operatorframework.io/) in the [OperatorHub.io](https://operatorhub.io/) or via vendors like Openshift this new helper plugin will help them out achieve further [Operator Capability levels](https://sdk.operatorframework.io/docs/overview/operator-capabilities/) since for Operator projects are able to achieve the ultimate levels Deep Insight and auto-pilot the must provide mechanisms to allow cluster admins Monitoring their solutions. 
```

### Goals

- Provide manifests that can display operator status for common needs.

> - Add a plugin or enhance the current kustomize plugin that scaffolds Grafana manifests.

Reviewer: Here we need to speak more about the business goals/generic and not how to achieve them. So, wdyt about something like:

- Provide an option to help Operator Authors provide monitoring features and projects with further quality which allow customer admins to check the exposed metrics with Grafana Dashboards 
- Encourage good practices and increase the quality of the Operator solutions distributed
- Provide good examples and reduce the TOLL for Operator Authors add value on their solutions as to skill up

### Non-Goals

> - TBD

Reviewer: ```suggestion
- Provided dashboards or helps for metrics which are not exported by default 
```
